src/libstd/sys/unix/stdio.rs

```diff
@@ -25,6 +39,21 @@ impl Stdin {
         fd.into_raw();
         ret
     }
+
+    pub fn close() -> io::Result<()> {
+        // To close stdin, what we actually do is change its well-known
+        // file descriptor number to refer to a file open on the null
+        // device.  This protects against code (perhaps in third-party
+        // libraries) that assumes STDIN_FILENO is always open and always
+        // refers to the same thing as stdin.
+        let mut fd = FileDesc::new(libc::STDIN_FILENO);
+        // If this step succeeds, the "previous" file descriptor returned
+        // by `fd.replace` is dropped and thus closed.
+        fd.replace(open_null_device(true)?)?;
+        // Don't close STDIN_FILENO itself, though!
+        fd.into_raw();
```

Comment:
Errors in those prior `?` will drop `fd` without calling your `into_raw`.  Perhaps you could wrap it in `ManuallyDrop` instead?

Reply:
Oh, good point.  Thanks for the tip!